### PR TITLE
25 add tag

### PR DESCRIPTION
**Pull request by oleynikovo**
Added add tag functionality, an activity screen for that, relevant tests.

Edited TagController into a singleton, fixed the equals() method in Tag class, added a couple functions to TagList class

### CODE REVIEW

File: ShinyExpenseTracker/src/ca/ualberta/cs/shinyexpensetracker/AddTagActivity.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import ca.ualberta.cs.shinyexpensetracker.R.menu;
+import android.app.Activity;
+import android.content.Intent;
+import android.os.Bundle;
+import android.view.Menu;
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.EditText;
+
+public class AddTagActivity extends Activity {
```
Comment (adamjford) [inline ShinyExpenseTracker/src/ca/ualberta/cs/shinyexpensetracker/AddTagActivity.java]: Need to have a comment here, as per Project Part 4 requirements:

>For each source code file, you should have a brief introductory comment describing its purpose or role within the application or a design pattern, as well as any currently outstanding issues.

File: ShinyExpenseTracker/src/ca/ualberta/cs/shinyexpensetracker/TagController.java [hunk 2]
```diff
@@ -1,21 +1,78 @@
...
+	 * @param s the string value of the added tag
+	 */
+	public static void addTag(String s) {
+		Pattern p = Pattern.compile("^[a-zA-Z0-9]*$");
+		Matcher m = p.matcher(s);
+		if (m.matches())
+			tagList.addTag(s);
+		else
+			return;
+	}
```
Comment (ItsTristan) [inline ShinyExpenseTracker/src/ca/ualberta/cs/shinyexpensetracker/TagController.java]: Minor details here:
* If the user enters an invalid string, is an error displayed?
* If the user enters an empty string, is that considered valid?
* If the user enters the same string a second time, is it added twice?

**Comment by ItsTristan** [comment]
Looks good. Some things to fix:
* Anywhere that you're using `tagList.doSomething(...)` in TagController, you should change it to `getTagList().doSomething(...)`. That should get rid of null pointers in the future.
* Your test for invalid tags doesn't cover a lot of edge cases (e.g. "word1 word2", "", "??symbol??", null, ...)
* You need tests for the UI.

File: ShinyExpenseTracker/src/ca/ualberta/cs/shinyexpensetracker/TagController.java (hunk 2 above)
Comment (oleynikovo) [inline ShinyExpenseTracker/src/ca/ualberta/cs/shinyexpensetracker/TagController.java]: * It now displays a Toast and asks for a new input
* Not anymore
* Isn't that kind of handled by ArrayList?

**Comment by oleynikovo** [comment]
Ok, how about now?
